design/ceph-device-set.md

```diff
@@ -0,0 +1,86 @@
+# Device Sets for rook-ceph
+
+**Target version**: Rook 1.1
+
+## Background
+The primary motivation for this feature is to take advantage of the mobility of storage in cloud-based environments. In environments like AWS you can request remote storage that can be dynamically provisioned and attached to any node in a given Availability Zone (AZ). If rook-ceph were able to take advantage of this it would improve the resiliency of the cluster and further automate disaster recovery. In addition, the design can also accommodate non-cloud environments through the use of local PVs.
+
+## CephDeviceSet CRD
+```yaml
+apiVersion: "ceph.rook.io/v1"
+kind: CephDeviceSet
+metadata:
+  name: my-device-set
+  namespace: rook-ceph
+spec:
+  # The Ceph cluster these devices belong to
+  cluster: my-ceph-cluster
+  # Number of devices in this set
+  count: 3
+  # Capacity of each device
+  capacity: 1Ti
+  # StorageClass to provision the storage devices
+  storageClassName: aws-ebs
+  # Resource requests/limits to apply to the OSD pods
+  resources:
+    requests:
+      cpu: 2000m
+      memory: 4Gi
+    limits:
+      cpu: 2000m
+      memory: 4Gi
+  # Placement constraints to apply to the OSD pods
+  placement:
+    nodeAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+        nodeSelectorTerms:
+        - matchExpressions:
+          - key: "failure-domain.beta.kubernetes.io/zone"
+            operator: In
+            values:
+            - us-west-1a
+    podAntiAffinity:
+      preferredDuringSchedulingIgnoredDuringExecution:
+      - weight: 100
+        podAffinityTerm:
+          labelSelector:
+            matchExpressions:
+            - key: "ocs.openshift.io/cluster"
+              operator: In
+              values:
+              - my-ceph-cluster
+          topologyKey: "failure-domain.beta.kubernetes.io/zone"
+```
+
+The CephDeviceSet CRD will define characteristics of a set of OSD devices and their Pods. The following fields are available in the spec:
```

Comment:
Instead of `CephDeviceSet` CRD is it possible to have `DeviceSet` CRD which will define characterstics of a set of devices which can be consumed by the storage providers. For example a DeviceSet has a count 3, Ceph can use it for backing 3 OSDs, NFS operator can use it for creating 3 exports, cassandra can use it for its 3 rack members(cassandra nodes)

Reply:
Also as per @travisn comment https://github.com/rook/rook/pull/3306#discussion_r294321699 if the pattern for OSD is not changed, this makes a good general functionality.